File: sgl-kernel/csrc/cpu/cpu_kunpeng/graph/capture.cpp
/*
 * Copyright 2026 Huawei Technologies Co., Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * ==============================================================================
 */

#include <algorithm>
#include <iostream>

#include "capture.h"

CaptureManager& CaptureManager::instance()
{
    static CaptureManager mgr;
    return mgr;
}

void CaptureManager::begin_capture(const std::vector<StorageBuf>& external_bufs,
                                   const std::vector<TensorView>& external_views,
                                   int num_inputs)
{
    capturing_ = true;
    num_external_inputs_ = num_inputs;
    storages_.clear();
    views_.clear();
    op_records_.clear();
    storage_registry_.clear();

    for (size_t i = 0; i < external_bufs.size(); ++i) {
        auto buf = external_bufs[i];
        buf.id = static_cast<int>(storages_.size());
        buf.born_op = -1;
        buf.death_op = -1;
        buf.in_pool = false;
        storage_registry_[buf.storage_base] = buf.id;
        storages_.push_back(buf);
    }

    for (size_t i = 0; i < external_views.size(); ++i) {
        auto view = external_views[i];
        view.id = static_cast<int>(views_.size());
        auto& s = storages_[view.storage_id];
        s.size = std::max(s.size,
            static_cast<size_t>(view.storage_offset) * view.element_size +
                view.numel * view.element_size);
        views_.push_back(view);
    }
}

std::pair<std::vector<StorageBuf>, std::vector<TensorView>>
CaptureManager::end_capture()
{
    TORCH_CHECK(capturing_, "end_capture: not in capture context");
    capturing_ = false;
    return {std::move(storages_), std::move(views_)};
}

int CaptureManager::lookup_storage(void* storage_base) const
{
    auto it = storage_registry_.find(storage_base);
    if (it != storage_registry_.end()) return it->second;
    return -1;
}

int CaptureManager::register_storage(StorageBuf buf)
{
    buf.id = static_cast<int>(storages_.size());
    storage_registry_[buf.storage_base] = buf.id;
    storages_.push_back(buf);
    return buf.id;
}

int CaptureManager::register_output_storage(StorageBuf buf)
{
    TORCH_CHECK(capturing_, "register_output_storage: not in capture context");
    buf.born_op = static_cast<int>(op_records_.size());
    return register_storage(buf);
}

int CaptureManager::register_view(TensorView view)
{
    TORCH_CHECK(capturing_, "register_view: not in capture context");
    view.id = static_cast<int>(views_.size());
    views_.push_back(view);
    return view.id;
}

int CaptureManager::find_or_register_view(
        int storage_id, int64_t storage_offset,
        const torch::Tensor& t)
{
    TORCH_CHECK(capturing_, "find_or_register_view: not in capture context");

    if (!t.defined()) {
        for (const auto& v : views_) {
            if (v.storage_id == storage_id && v.storage_offset == 0 &&
                v.numel == 0 && v.element_size == 0 &&
                v.shape.empty() && !v.is_return) {
                return v.id;
            }
        }
        TORCH_CHECK(storage_id >= num_external_inputs_,
            "input tensor (storage=", storage_id, ") cannot be undefined/None");
        TensorView view;
        view.storage_id = storage_id;
        view.storage_offset = 0;
        view.numel = 0;
        view.element_size = 0;
        view.scalar_type = 0;
        view.is_return = false;
        return register_view(view);
    }

    size_t numel = static_cast<size_t>(t.numel());
    size_t element_size = static_cast<size_t>(t.element_size());
    int scalar_type = static_cast<int>(t.scalar_type());
    std::vector<int64_t> shape(t.sizes().begin(), t.sizes().end());
    std::vector<int64_t> strides(t.strides().begin(), t.strides().end());

    for (const auto& v : views_) {
        if (v.storage_id == storage_id &&
            v.storage_offset == storage_offset &&
            v.numel == numel &&
            v.element_size == element_size &&
            v.scalar_type == scalar_type &&
            v.shape == shape &&
            v.strides == strides &&
            !v.is_return) {
            return v.id;
        }
    }
    // No matching view. For input storages: must match the pre-registered view.
    TORCH_CHECK(storage_id >= num_external_inputs_,
        "input tensor (storage=", storage_id, ") has no matching view");
    TensorView view;
    view.storage_id = storage_id;
    view.storage_offset = storage_offset;
    view.numel = numel;
    view.element_size = element_size;
    view.scalar_type = scalar_type;
    view.shape = shape;
    view.strides = strides;
    view.is_return = false;
    return register_view(view);
}

void CaptureManager::record_op(OpRecord op)
{
    TORCH_CHECK(capturing_, "record_op: not in capture context");
    if constexpr (kGraphDebugPrint)
        debug_print_op(op);
    op_records_.push_back(std::move(op));
}

void CaptureManager::debug_print_op(const OpRecord& op) const
{
    auto print_view = [this](int vid, const char* tag) {
        const auto& v = views_[vid];
        int sid = v.storage_id;
        const auto& s = storages_[sid];
        const char* source;
        if (s.born_op < 0) {
            source = (vid < num_external_inputs_) ? "input" : "fixed";
        } else {
            source = "intermediate";
        }
        std::cout << "  " << tag << " vid=" << vid << " sid=" << sid
                  << " source=" << source << " shape=[";
        for (size_t i = 0; i < v.shape.size(); ++i) {
            if (i) std::cout << ", ";
            std::cout << v.shape[i];
        }
        std::cout << "]" << std::endl;
    };

    int op_idx = static_cast<int>(op_records_.size());
    std::cout << "[op " << op_idx << "] " << op.op_name << std::endl;
    for (int vid : op.input_view_ids)  print_view(vid, " in ");
    for (int vid : op.output_view_ids) print_view(vid, " out");
    std::cout << std::endl;
}

GraphOpRegistry& GraphOpRegistry::instance()
{
    static GraphOpRegistry reg;
    return reg;
}

void GraphOpRegistry::register_op(const std::string& name, OpInfo info)
{
    ops_[name] = std::move(info);
}

const GraphOpRegistry::OpInfo* GraphOpRegistry::lookup(const std::string& name) const
{
    auto it = ops_.find(name);
    if (it != ops_.end()) return &it->second;
    return nullptr;
}

File: python/sglang/srt/graph/_capture.py
import torch

from sgl_kernel import graph_cpp as _C

_captured_storages = None
_captured_views = None
_captured_ops = None
_captured_num_inputs = None
_captured_fixed = None
_none_storage_id = None
_captured_input_map = None


class _CaptureContext:
    def __init__(self, inputs, fixed=None):
        self.inputs = inputs
        self.fixed = fixed or []

    def __enter__(self):
        global _captured_storages, _captured_views, _captured_ops, \
               _captured_num_inputs, _captured_fixed, _none_storage_id, \
               _captured_input_map
        _captured_storages = None
        _captured_views = None
        _captured_ops = None
        _captured_num_inputs = None
        _captured_fixed = None
        _none_storage_id = None
        _captured_input_map = None
        mgr = _C.CaptureManager.instance()
        mgr.begin_capture(self.inputs, self.fixed)
        _captured_input_map = {}
        for i, t in enumerate(self.inputs):
            _captured_input_map[id(t)] = (i, _C.storage_offset(t), i)
        return self

    def __exit__(self, exc_type, exc_val, exc_tb):
        global _captured_storages, _captured_views, _captured_ops, \
               _captured_num_inputs, _captured_fixed, _captured_input_map
        mgr = _C.CaptureManager.instance()
        if mgr.is_capturing():
            _captured_storages, _captured_views, _captured_ops, \
                _captured_num_inputs = mgr.end_capture()
            n = len(self.inputs)
            _captured_fixed = {n + i: t for i, t in enumerate(self.fixed)}
        _captured_input_map = None
        return False


def capture(inputs, fixed=None):
    return _CaptureContext(inputs, fixed)


def finalize(outputs, external_pool=None, external_shm_pool=None):
    global _captured_storages, _captured_views, _captured_ops, \
           _captured_num_inputs, _captured_fixed, _captured_input_map
    if _captured_storages is None:
        raise RuntimeError("No captured graph to finalize")
    storages = _captured_storages
    views = _captured_views
    ops = _captured_ops
    num_inputs = _captured_num_inputs
    fixed = _captured_fixed or {}
    _captured_storages = None
    _captured_views = None
    _captured_ops = None
    _captured_num_inputs = None
    _captured_fixed = None
    _captured_input_map = None

    output_view_ids = []
    mgr = _C.CaptureManager.instance()
    for t in outputs:
        base = _C.storage_base(t)
        so = _C.storage_offset(t)
        sid = mgr.lookup_storage(base)
        if sid < 0:
            raise RuntimeError(
                f"Output tensor (base {base}) not found in any storage")

        if sid < num_inputs:
            # Output on input storage: exact match with existing view
            t_shape = list(t.shape)
            t_strides = list(t.stride())
            found = None
            for view in views:
                if (view.storage_id == sid and
                    view.storage_offset == so and
                    view.shape == t_shape and
                    view.strides == t_strides and
                    not view.is_return):
                    found = view
                    break
            if found is None:
                raise RuntimeError(
                    f"Output tensor (sid {sid}) has no matching input view")
            output_view_ids.append(found.id)
        else:
            # Output on fixed or intermediate storage: always create new view
            _, view = _C.tensor_to_buf_and_view(t)
            view.storage_id = sid
            view.is_return = True
            view.id = len(views)
            views.append(view)
            output_view_ids.append(view.id)

    gh = _C.Graph(storages, views, ops, output_view_ids, num_inputs, fixed,
                  external_pool, external_shm_pool)
    return gh


def is_capturing():
    return _C.CaptureManager.instance().is_capturing()


def lookup_or_register(tensor, idx=0):
    global _none_storage_id, _captured_input_map
    mgr = _C.CaptureManager.instance()
    if tensor is None:
        if _none_storage_id is None:
            buf = _C.StorageBuf()
            buf.storage_base = 0
            buf.size = 0
            buf.born_op = 0
            _none_storage_id = mgr.register_storage(buf)
        vid = mgr.find_or_register_view(_none_storage_id, 0, tensor)
        return vid, 0, _none_storage_id
    if _captured_input_map is not None:
        cached = _captured_input_map.get(id(tensor))
        if cached is not None:
            return cached
    base = _C.storage_base(tensor)
    assert base != 0, (
        f"non-input tensor has nullptr storage at {idx}-th parameter: "
        f"shape={tuple(tensor.shape)}, dtype={tensor.dtype}"
    )
    so = _C.storage_offset(tensor)
    sid = mgr.lookup_storage(base)
    if sid < 0:
        print(
            f"[capture] lookup_or_register FAIL at {idx}-th parameter: "
            f"shape={tuple(tensor.shape)}, dtype={tensor.dtype}, "
            f"numel={tensor.numel()}, "
            f"data_ptr=0x{tensor.data_ptr():x}, "
            f"storage_base=0x{base:x}",
            flush=True,
        )
    assert sid >= 0, "non-return-value parameter tensor not registered"

    vid = mgr.find_or_register_view(sid, so, tensor)
    return vid, so, sid


_MEMORY_TYPE_TO_ENUM = {
    'regular': _C.MemoryType.REGULAR,
    'shm': _C.MemoryType.SHM,
}


def register_output(tensor, memory_type='regular'):
    mgr = _C.CaptureManager.instance()
    so = _C.storage_offset(tensor)
    buf, view = _C.tensor_to_buf_and_view(tensor)
    sid = mgr.register_output_storage(buf, _MEMORY_TYPE_TO_ENUM[memory_type])
    view.storage_id = sid
    view.is_return = True
    vid = mgr.register_view(view)
    return vid, so


def upgrade_storage_memory_type(tensor):
    """Upgrade the storage backing ``tensor`` to SHM memory type."""
    mgr = _C.CaptureManager.instance()
    base = _C.storage_base(tensor)
    sid = mgr.lookup_storage(base)
    assert sid >= 0, (
        f"upgrade_storage_memory_type: storage not registered, "
        f"shape={tuple(tensor.shape)}, dtype={tensor.dtype}"
    )
    mgr.upgrade_storage_memory_type(sid)


def record_op(op_name, inputs, outputs, scalar_args, profile_name=""):
    mgr = _C.CaptureManager.instance()
    op = _C.OpRecord()
    op.op_name = op_name
    op.profile_name = profile_name
    op.input_view_ids = [vid for vid, _, _ in inputs]
    op.output_view_ids = [vid for vid, _ in outputs]
    op.scalar_args = list(scalar_args)
    mgr.record_op(op)
